File: PHP/Solid Principles/open_close/solution/PaymentProcessor.php
<?php

use Solid\Interfaces\PaymentProcessorInterface;
use Solid\HelperClasses\Payment;
use Solid\HelperClasses\Order;
use Solid\Processors\VisaProcessor;
use Solid\Processors\AmericanExpressProcessor;
use Solid\Processors\MasterCardProcessor;

class PaymentProcessor
{

    public function processPayment(Order $order, Payment $payment)
    {
        echo "Processing payment of order {$order->getName()}\n";
        echo "Issuing payment for amount {$order->getPrice()}\n";

        // Get the processor based on payment type
        $processor = $this->getPaymentProcessor($payment->getType());

        // Process the payment based on the correct processor
        $processor->process($order, $payment);
    }

    private function getPaymentProcessor(string $type): PaymentProcessorInterface
    {
        switch ($type) {
            case 'VISA':
                return new VisaProcessor();
            case 'Master_Card':
                return new MasterCardProcessor();
            case 'American_Express':
                return new AmericanExpressProcessor();
            default:
                throw new Exception("Unsupported payment type: {$type}");
        }
    }
}

File: Python/SOLID/Single-Responsibility/PaymentProcessor.py
from Payment import Payment


class PaymentProcessor:
    def process_payment(self, order : Order, payment : Payment) -> None:
        # process payment
        print("Processing payment for order: ", order.get_order_name())
        print("Issuing payment for amount: ", order.get_order_price())

        # check payment type
        if payment.get_type() == "VISA":
            print("Processing VISA Card payment")
        elif payment.get_type() == "MASTERCARD":
            print("Processing MASTERCARD payment")
        elif payment.get_type() == "PAYPAL":
            print("Processing PAYPAL payment")
        else:
            print("Invalid payment type")

